In [110]:
# Machine Learning


In [111]:
# independent = x , dependent = y

In [112]:
df

Time_spent_Alone  Stage_fear  Social_event_attendance  Going_outside  \
id                                                                           
0                  4.0           0                      4.0            6.0   
1                  9.0           1                      0.0            0.0   
4                  3.0           0                      9.0            4.0   
5                  1.0           0                      7.0            5.0   
6                  4.0           0                      9.0            3.0   
...                ...         ...                      ...            ...   
2895               3.0           0                      7.0            6.0   
2896               3.0           0                      8.0            3.0   
2897               4.0           1                      1.0            1.0   
2898              11.0           1                      1.0            3.0   
2899               3.0           0                      6.0            6.0   

      Drained_after_socializing  Friends_circle_size  Post_frequency  \
id                                                                     
0                             0                 13.0             5.0   
1                             1                  0.0             3.0   
4                             0                  8.0             5.0   
5                             0                  6.0             6.0   
6                             0                  7.0             7.0   
...                         ...                  ...             ...   
2895                          0                  6.0             6.0   
2896                          0                 14.0             9.0   
2897                          1                  4.0             0.0   
2898                          1                  2.0             0.0   
2899                          0                  6.0             9.0   

      Personality  
id                 
0               0  
1               1  
4               0  
5               0  
6               0  
...           ...  
2895            0  
2896            0  
2897            1  
2898            1  
2899            0  

[2498 rows x 8 columns]

In [113]:
df.columns

Index(['Time_spent_Alone', 'Stage_fear', 'Social_event_attendance',
       'Going_outside', 'Drained_after_socializing', 'Friends_circle_size',
       'Post_frequency', 'Personality'],
      dtype='object')

In [114]:
x = df.iloc[:,:-1] # x = defining independent columns with all the row,
y = df.iloc[:,-1]  # y = defining dependent columns with all the row,

In [115]:
x.shape,y.shape

((2498, 7), (2498,))

In [116]:
x.columns

Index(['Time_spent_Alone', 'Stage_fear', 'Social_event_attendance',
       'Going_outside', 'Drained_after_socializing', 'Friends_circle_size',
       'Post_frequency'],
      dtype='object')

In [117]:
y

id
0       0
1       1
4       0
5       0
6       0
       ..
2895    0
2896    0
2897    1
2898    1
2899    0
Name: Personality, Length: 2498, dtype: int64

In [118]:
df["Personality"].value_counts()

Personality
0    1403
1    1095
Name: count, dtype: int64

# **Data epo kudhalum -> Train and test ku separate pannanum or train_test_split | For the following steps has to be considered**

# 2498 - **orginal df row size**
# **train** - 70% == 0.7*2498 = 1749
# **test** - max = 30% == 0.3*2498 = 749

# **Personality**                      
# 0	1403  **Extrovert**             |  1 - 	1095  - **Introvert**
# 30% for **test** = 421 &        |  30% for **test** = 329
# 70% for **train** = 982         |  70% for **train** = 767

# 421+329 -> 750 --> **Test**
# 982+767 -> 1749 --> **Train**

In [127]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size = 0.3)
x_train.shape,x_test.shape,y_train.shape,y_test.shape

((1748, 7), (750, 7), (1748,), (750,))

In [128]:
y_train.value_counts()  #982+767

Personality
0    964
1    784
Name: count, dtype: int64

In [129]:
y_test.value_counts() #421+329

Personality
0    439
1    311
Name: count, dtype: int64

# **Types of Classification Algorithms:**
**Logistic Regression:**
A linear model, often used for binary classification tasks (e.g., yes/no, true/false). It predicts the probability of a data point belonging to a specific class by fitting a logistic function to the data.
**Decision Trees:**
Non-linear models that use a tree-like structure to make decisions. They recursively split the data based on feature values to classify data points.
**Random Forests:**
An ensemble method that combines multiple decision trees to improve accuracy and reduce overfitting.
**Support Vector Machines (SVM):**
A powerful algorithm that finds the optimal hyperplane to separate data points into different classes, particularly effective for high-dimensional data.
**K-Nearest Neighbors (KNN):**
A simple, instance-based algorithm that classifies data points based on the majority class of their k-nearest neighbors in the feature space.
**Naive Bayes:**
A probabilistic classifier that applies Bayes' theorem with strong (naive) independence assumptions between features.
**Neural Networks:**
Complex models with interconnected layers of artificial neurons, capable of learning intricate patterns and relationships in data, suitable for tasks like image recognition.

In [132]:
from sklearn.tree import DecisionTreeClassifier

In [133]:
model= DecisionTreeClassifier().fit(x_train,y_train) # ((67302, 8) & # 5937+61365

In [134]:
import numpy as np


In [136]:
x

Time_spent_Alone  Stage_fear  Social_event_attendance  Going_outside  \
id                                                                           
0                  4.0           0                      4.0            6.0   
1                  9.0           1                      0.0            0.0   
4                  3.0           0                      9.0            4.0   
5                  1.0           0                      7.0            5.0   
6                  4.0           0                      9.0            3.0   
...                ...         ...                      ...            ...   
2895               3.0           0                      7.0            6.0   
2896               3.0           0                      8.0            3.0   
2897               4.0           1                      1.0            1.0   
2898              11.0           1                      1.0            3.0   
2899               3.0           0                      6.0            6.0   

      Drained_after_socializing  Friends_circle_size  Post_frequency  
id                                                                    
0                             0                 13.0             5.0  
1                             1                  0.0             3.0  
4                             0                  8.0             5.0  
5                             0                  6.0             6.0  
6                             0                  7.0             7.0  
...                         ...                  ...             ...  
2895                          0                  6.0             6.0  
2896                          0                 14.0             9.0  
2897                          1                  4.0             0.0  
2898                          1                  2.0             0.0  
2899                          0                  6.0             9.0  

[2498 rows x 7 columns]

In [140]:
vp= np.array([11.0,1.0,7.0,5.0,1.0,7.0,6.0])

In [142]:
model.predict(vp.reshape(1,-1))

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning:

X does not have valid feature names, but DecisionTreeClassifier was fitted with feature names



array([1])

In [143]:
y_pred = model.predict(x_test)
y_pred

array([0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0,
       0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0,
       0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0,
       0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1,
       1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0,
       1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1,
       0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0,
       0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0,

In [144]:
y_test

id
2220    0
1291    1
2255    1
1979    1
171     0
       ..
126     0
1125    0
2553    0
1327    0
174     0
Name: Personality, Length: 750, dtype: int64

In [145]:
from sklearn.metrics import confusion_matrix, accuracy_score,classification_report

In [146]:
confusion_matrix(y_test,y_pred)

array([[377,  62],
       [ 60, 251]])

In [147]:
accuracy_score(y_test,y_pred)

0.8373333333333334

In [148]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.86      0.86      0.86       439
           1       0.80      0.81      0.80       311

    accuracy                           0.84       750
   macro avg       0.83      0.83      0.83       750
weighted avg       0.84      0.84      0.84       750

